tcp_sink.go

```diff
@@ -4,105 +4,128 @@ import (
 	"bufio"
 	"fmt"
 	"net"
-	"sync/atomic"
+	"sync"
 	"time"
 
 	logger "github.com/sirupsen/logrus"
 )
 
 const (
-	logOnEveryNDropped = 1000
+	flushInterval           = time.Second
+	logOnEveryNDroppedBytes = 1 << 15 // Log once per 32kb of dropped stats
 )
 
-// NewTCPStatsdSink returns a Sink that is backed by a go channel with a limit of 1000 messages.
+// NewTCPStatsdSink returns a Sink that is backed by a buffered writer and a separate goroutine
+// that flushes those buffers to a statsd connection.
 func NewTCPStatsdSink() Sink {
+	outc := make(chan []byte, 1000)
 	sink := &tcpStatsdSink{
-		outc: make(chan string, 1000),
+		outc:     outc,
+		flushBuf: bufio.NewWriter(sinkWriter{outc: outc}),
 	}
 	go sink.run()
 	return sink
 }
 
 type tcpStatsdSink struct {
-	conn            net.Conn
-	outc            chan string
-	droppedTimers   uint64
-	droppedCounters uint64
-	droppedGauges   uint64
+	conn         net.Conn
+	outc         chan []byte
+	droppedBytes uint64
+	mu           sync.Mutex
+	flushBuf     *bufio.Writer
 }
 
-func (s *tcpStatsdSink) FlushCounter(name string, value uint64) {
+type sinkWriter struct {
+	outc chan []byte
+}
+
+func (w sinkWriter) Write(p []byte) (int, error) {
+	n := len(p)
+	pCopy := make([]byte, n)
+	copy(pCopy, p)
```

Comment:
why copy?

Reply:
isbufio going to reuse that buffer? not part of the contract afaict

Reply:
https://golang.org/pkg/io/#Writer
"Implementations must not retain p."

Reply:
ah. i guess we are trading off (a) not blocking the emitter goroutine for (b) an intermediate copy

Reply:
can you make that a reusable buffer from a buffer pool? put it back in the pool after writing it to the conn

Reply:
and fetch from the pool before the copy

Reply:
sync.Pool